lib/ZnapZend/ZFS.pm

```diff
@@ -431,28 +456,129 @@ sub sendRecvSnapshots {
+    # the process, and/or time out. To defend against this, we optionally
+    # can fall back to a big list of individual dataset names found by
+    # recursive listDataSets() invocations instead.
+    if (defined($dataSet) && $dataSet) {
+        if (ref($dataSet) eq 'ARRAY') {
+            print STDERR "=== getDataSetProperties(): Is array...\n" if $self->debug;
+            if ($self->lowmemRecurse && $recurse) {
+                my $listds = $self->listDataSets(undef, $dataSet, $recurse);
+                if (scalar(@{$listds}) == 0) {
+                    print STDERR "=== getDataSetProperties(): Failed to get data from listDataSets()\n" if $self->debug;
+                    return \@propertyList;
+                }
+                push (@list, @{$listds});
+                $recurse = 0;
+            } else {
+                if ( (scalar(@$dataSet) > 0) && (defined(@$dataSet[0])) ) {
+                    push (@list, @$dataSet);
+                } else {
+                    print STDERR "=== getDataSetProperties(): skip array context: value(s) inside undef...\n" if $self->debug;
+                }
+            }
+        } else {
+            # Assume a string, per usual invokation
+            print STDERR "=== getDataSetProperties(): Is string...\n" if $self->debug;
+            if ($self->lowmemRecurse && $recurse) {
+                my $listds = $self->listDataSets(undef, $dataSet, $recurse);
+                if (scalar(@{$listds}) == 0) {
+                    print STDERR "=== getDataSetProperties(): Failed to get data from listDataSets()\n" if $self->debug;
+                    return \@propertyList;
+                }
+                push (@list, @{$listds});
+                $recurse = 0;
+            } else {
+                if ($dataSet ne '') {
+                    push (@list, ($dataSet));
+                } else {
+                    print STDERR "=== getDataSetProperties(): skip string context: value(s) inside undef...\n" if $self->debug;
+                }
+            }
+        }
+    } else {
+        print STDERR "=== getDataSetProperties(): no dataSet argument passed\n" if $self->debug;
+    }
+
+    if (scalar(@list) == 0) {
+        print STDERR "=== getDataSetProperties(): List all local datasets on the system...\n" if $self->debug;
+        my $listds = $self->listDataSets();
+        if (scalar(@{$listds}) == 0) {
+            print STDERR "=== getDataSetProperties(): Failed to get data from listDataSets()\n" if $self->debug;
+            return \@propertyList;
+        }
+        push (@list, @{$listds});
+    }
+
+    for my $listElem (@list){
+        print STDERR "=== getDataSetProperties(): Looking under '$listElem' with '$recurse' recursion mode\n" if $self->debug;
         my %properties;
-        my @cmd = (@{$self->priv}, qw(zfs get -H -s local -o), 'property,value', 'all', $listElem);
+        my @cmd = (@{$self->priv}, qw(zfs get -H -s local));
+        push (@cmd, qw(-t), 'filesystem,volume');
```

Comment:
Also for reviewers: did any users of the older implementation of `getDataSetProperties()` rely on its ability to query non-fs/vol datasets (e.g. ask for properties of snapshots, or of whatever future dataset types may be defined)?

It can be added back for non-recursive case of explicitly listed dataset(s) to query, but it explodes size- and time-wise if we recurse a pool with hundreds of thousands of snapshots and get attributes of those too.

Reply:
I am not aware of such a use ...